hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/TestHoodieSparkSqlWriter.scala

```diff
@@ -960,6 +960,86 @@ class TestHoodieSparkSqlWriter {
     assert(spark.read.format("hudi").load(tempBasePath).where("age >= 2000").count() == 10)
   }
 
+  /**
+   * Test upsert for CoW table without precombine field and combine before upsert disabled.
+   */
+  @Test
+  def testUpsertWithoutPrecombineFieldAndCombineBeforeUpsertDisabled(): Unit = {
+    val options = Map(DataSourceWriteOptions.TABLE_TYPE.key -> HoodieTableType.COPY_ON_WRITE.name(),
```

Comment:
only for CoW since I assume MoR requires pecombine field

Reply:
@kazdy : Can you also cover MOR case. Even for MOR, we should let upsert skip pre-combine if the user expects input batch to be unique and wants to skip this step.